rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,385 @@
+   (`service-account.yaml`, `role.yaml`, `role-binding.yaml`, `configmap.yaml`).
+   Which approach is better? Why are they different?
+1. Should we prefer `kubectl create ...` or `kubectl apply ...`? The examples
+   in the [prior art](#prior-art) section use both.
+1. From what I understand, Vector requires the Kubernetes `watch` verb in order
+   to receive updates to k8s cluster changes. This is required for the
+   `kubernetes_pod_metadata` transform. Yet, Fluentbit [requires the `get`,
+   `list`, and `watch` verbs][fluentbit_role]. Why don't we require the same?
+1. What is `updateStrategy` ... `RollingUpdate`? This is not included in
+   [our daemonset][vector_daemonset] or in [any of Fluentbit's config
+   files][fluentbit_installation]. But it is included in both [Fluentd's
+   daemonset][fluentd_daemonset] and [LogDNA's daemonset][logdna_daemonset].
+1. I've also noticed `resources` declarations in some of these config files.
+   For example [LogDNA's daemonset][logdna_daemonset]. I assume this is limiting
+   resources. Do we want to consider this?
+1. What the hell is going on with [Honeycomb's integration
+   strategy][honeycomb integration]? :) It seems like the whole "Heapster"
+   pipeline is specifically for system events, but Heapster is deprecated?
+   This leads me to my next question...
+1. How are we collecting Kubernetes system events? Is that outside of the
+   scope of this RFC? And why does this take an entirely different path?
+   (ref [issue#1293])
+1. What are some of the details that set Vector's Kubernetes integration apart?
+   This is for marketing purposes and also helps us "raise the bar".
+
+### From Mike
+
+1. What significantly different k8s cluster "flavors" are there? Which ones do
+   we want to test against? Some clusters use `docker`, some use `CRI-O`,
+   [etc][container_runtimes]. Some even use [gVisor] or [Firecracker]. There
+   might be differences in how different container runtimes handle logs.
+
+## Plan Of Attack
+
+- [ ] Agree on minimal Kubernetes version.
+- [ ] Agree on a list of Kubernetes cluster flavors we want to test against.
+- [ ] Setup a proper testing suite for k8s.
+  - [ ] Support for customizable k8s clusters. See [issue#2170].
+  - [ ] Look into [issue#2225] and see if we can include it as part of this
+        work.
+  - [ ] Stabilize k8s integration tests. See [issue#2193], [issue#2216],
+        and [issue#1635].
+  - [ ] Ensure we are testing all supported minor versions. See
+        [issue#2223].
+- [ ] Audit and improve the `kubernetes` source.
+  - [ ] Handle the log recursion problem where Vector ingests it's own logs.
+        See [issue#2218] and [issue#2171].
+  - [ ] Audit the `file` source strategy. See [issue#2199] and [issue#1910].
+    - [ ] Merge split logs. See [pr#2134].
+- [ ] Audit and improve the `kubernetes_pod_matadata` transform.
+  - [ ] Use the `log_schema.kubernetes_key` setting. See [issue#1867].
+- [ ] Ensure our config reload strategy is solid.
```

Comment:
While k8s does update mounted config maps on change, it's has a certain delay: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically.
I'd rather *explicitly disable* automatic config file reloads in the config by default (we'd have to add a key to allow that), and advice our users to [`kubectl rollut restart deployment/vector`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#-em-restart-em-). This gives more control and less surprises, and therefore, is a better default for k8s.

Reply:
Makes sense to me.

Reply:
Addressed this at the RFC itself